blog/_posts/2022-05-30-long-term-compatibility-plans.md

```diff
@@ -0,0 +1,85 @@
+
+More importantly, during work on 3.2, we realized that maintaining this flag may be more challenging than anticipated. With time, it will be increasingly harder to be sure that our handling of it is correct. **Thus we have decided to remove the possibility of configuring the output version altogether in Scala 3.2.**
+
+## Our solution: Scala LTS & Scala Next
+
+Finally, we have reached the agreement that the best course of action would be to split Scala development into two lines, called “Scala LTS” and “Scala Next”. Both would live as separate branches in the compiler repository. Both will have separate, but possibly synchronized releases. Porting changes from one to the other should be easy and common.
+
+Scala Next will be the line that the development of the language is taking place on. It would receive frequent minor updates and all experimental language features will live here.
+
+Scala LTS on the other hand will be the stable long-term support line. One LTS will only receive patch updates which means that all releases of the same LTS line will be forward and backward compatible in terms of output. The only changes between releases will be bug fixes, non-language changes (doctool, semanticDB, reporting), and small quality-of-life enhancements (only if we are sure we are not breaking any compatibility guarantees. After more than two years we may decide to nominate one of the releases from the Next line as the new LTS. After that, we guarantee that the old LTS will be supported for at least one additional year. That means that the time of support for each LTS release is at least three years.
+
+![Scala LTS Roadmap]({{ site.baseurl }}/resources/img/scala3-lts-roadmap.svg)
+
+### The compatibility guarantees in the new model
+
+The new model doesn’t change our compatibility guarantees. We are keeping forward and backward compatibility within a single minor version and backward compatibility across different minor versions. That means that the code compiled with Scala LTS can be used as a dependency by all Scala Next versions newer than said LTS, but not the other way around.
+
+## The (future) best practices
+
+Once the LTS & Next model is implemented, we suggest the following practices for different groups of users:
+
+### Owners of commercial projects
+
+You should stick to the LTS version. It will give you the best tooling support and a stable compiler. If you are migrating a bigger project from Scala 2, you should migrate it straight to the newest LTS.
```

Comment:
Could you explain more what the risks for bigger or commercial projects would be if they choose not to stick with an LTS release?

Reply:
IMO this is worded a bit too strongly. I think there are two valid choices for commercial projects:

* Stick to the LTS version -> update less often (from LTS to LTS), with potentially more changes to deal with at once, or
* Be up-to-date to the latest minor version -> update more often, but with smaller deltas

In both cases, you will get the best tooling support and a stable compiler.

The catch is that you take the option of the latest minor version, you will *have to* update more often in order to keep having the best tooling support, performance improvements and bug fixes (including potential security fixes).

If you are prepared to keep up in that sense, it is perfectly fine to use the latest version in a commercial project.

Reply:
> The catch is that you take the option of the latest minor version, you will have to update more often in order to keep having the best tooling support, performance improvements and bug fixes (including potential security fixes).

I suppose this was the only option before LTS releases were introduced?

Reply:
> If you are prepared to keep up in that sense, it is perfectly fine to use the latest version in a commercial project.

Yes that is indeed important information so that commercial projects, are not "scared away" from being up-to-date if they care about frequent updates. It is good to know that following the frequent release train also provides a stable compiler. (As I understand it, it's the RC:s that are less stable, and nightly even less stable, in terms of how much real-world testing has been made so far.)

Reply:
> I suppose this was the only option before LTS releases were introduced?

Yes, indeed.

> As I understand it, it's the RC:s that are less stable, and nightly even less stable, in terms of how much real-world testing has been made so far.

Also yes, indeed. ;)

Reply:
> * Be up-to-date to the latest minor version -> update more often, but with smaller deltas

This is impossible for the large majority of the codebases in the industry. Nevertheless, it is true. I will add a sentence about it being possible if the code maintainers are willing to commit to the constant updates.

Reply:
``` md
You may also choose to be up to date with the latest minor version. That will allow you to use new features as soon as they are added to the language. However, you will need to be ready for frequent compiler updates. Otherwise, your project may end up with an unsupported version of the language.
```
was added.

Reply:
I guess you want to update this blog as well, and remove some stuff:
https://www.scala-lang.org/blog/2022/04/12/scala-3.1.2-released.html
if scalaOutputVersion is cancelled, as well as sbt docs etc...
Otherwise there is a risk that people get hold of this and miss the updated policy on LTS etc.

Reply:
`-scala-output-version` was supported only in one pre-release version of the sbt, and it is being removed right now. Regarding the previous blog post, it may indeed be a good idea to add an update there.